satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,307 @@
+
+from satpy import Scene
+from satpy.modifiers import ModifierBase
+from satpy.utils import get_satpos, lonlat2xyz, xyz2lonlat
+
+
+class MissingHeightError(ValueError):
+    """Raised when heights do not overlap with area to be corrected."""
+
+
+class IncompleteHeightWarning(UserWarning):
+    """Raised when heights only partially overlap with area to be corrected."""
+
+
+def forward_parallax(sat_lon, sat_lat, sat_alt, lon, lat, height):
+    """Calculate forward parallax effect.
+
+    Calculate the forward parallax effect.  When a satellite instrument
+    observes the Earth, the geolocation assumes it sees the Earth surface at
+    the geoid (elevation zero).  In reality, the ray may stop short of the
+    geoid as it observes, for example, a cloud or elevated ground.  This
+    function calculates the forward parallax effect.  If the view of a pixel at
+    location (lat, lon) is blocked by a cloud at height h, we calculate the
+    location of this blocking.
+
+    Calculate parallax correction based on satellite position and
+    (cloud top) height coordinates in geodetic (unprojected) coordinates.
+    This function calculates the latitude and longitude belonging to the
+    cloud top, based on the location of the satellite and the location
+    of the cloud.
+
+    For scenes that are only partly cloudy, the user might set the cloud top
+    height for clear-sky pixels to NaN.  This function will return a corrected
+    lat/lon as NaN as well.  The user can use the original lat/lon for those
+    pixels or use the higher level :class:`ParallaxCorrection` class.
+
+    This function assumes a spherical Earth.
+
+    Args:
+        sat_lon (number): Satellite longitude in geodetic coordinates [°]
+        sat_lat (number): Satellite latitude in geodetic coordinates [°]
+        sat_alt (number): Satellite altitude above the Earth surface [km]
+        lon (array or number): Longitudes of pixel or pixels to be corrected,
+            in geodetic coordinates [°]
+        lat (array or number): Latitudes of pixel/pixels to be corrected, in
+            geodetic coordinates [°]
+        height (array or number): Heights of pixels on which the correction
+            will be based.  Typically this is the cloud based height. [km]
+
+    Returns:
+        tuple[float, float]: New geolocation
+            New geolocation ``(lon, lat)`` for the longitude and
+            latitude that were to be corrected, in geodetic coordinates. [°]
+    """
+    X_sat = np.hstack(lonlat2xyz(sat_lon, sat_lat)) * sat_alt
+    X = np.stack(lonlat2xyz(lon, lat), axis=-1) * EARTH_RADIUS
+    # the datetime doesn't actually affect the result but is required
+    # so we use a placeholder
+    (_, elevation) = get_observer_look(
+            sat_lon, sat_lat, sat_alt,
+            datetime(2000, 1, 1), lon, lat, EARTH_RADIUS)
+    # TODO: handle cases where this could divide by 0
+    parallax_distance = height / np.sin(np.deg2rad(elevation))
```

Comment:
Instead of a TODO, could we catch the division by zero error and raise a not implemented error?

Reply:
I'll do that.  It would divide by zero if the satellite is exactly at the horizon, I think.